File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/model/PeerConnectionState.kt
package com.instrumentisto.medea_flutter_webrtc.model

import org.webrtc.PeerConnection.PeerConnectionState as WPeerConnectionState

/**
 * Representation of an [org.webrtc.PeerConnection.PeerConnectionState].
 *
 * @property value [Int] representation of this enum which will be expected on the Flutter side.
 */
enum class PeerConnectionState(val value: Int) {
  /**
   * Indicates that any of the ICE transports or DTLS transports are in the "new" state and none of
   * the transports are in the "connecting", "checking", "failed" or "disconnected" state, or all
   * transports are in the "closed" state, or there are no transports.
   */
  NEW(0),

  /**
   * Indicates that any of the ICE transports or DTLS transports are in the "connecting" or
   * "checking" state and none of them is in the "failed" state.
   */
  CONNECTING(1),

  /**
   * Indicates that all the ICE transports and DTLS transports are in the "connected", "completed"
   * or "closed" state, and at least one of them is in the "connected" or "completed" state.
   */
  CONNECTED(2),

  /**
   * Indicates that any of the ICE transports or DTLS transports are in the "disconnected" state,
   * and none of them are in the "failed", "connecting" or "checking" state.
   */
  DISCONNECTED(3),

  /** Indicates that any of the ICE transports or DTLS transports are in the "failed" state. */
  FAILED(4),

  /** Indicates that the peer connection is closed. */
  CLOSED(5);

  companion object {
    /**
     * Converts the provided [org.webrtc.PeerConnection.PeerConnectionState] into a
     * [PeerConnectionState].
     *
     * @return [PeerConnectionState] created based on the provided
     * [org.webrtc.PeerConnection.PeerConnectionState].
     */
    fun fromWebRtc(from: WPeerConnectionState): PeerConnectionState {
      return when (from) {
        WPeerConnectionState.NEW -> NEW
        WPeerConnectionState.CONNECTING -> CONNECTING
        WPeerConnectionState.CONNECTED -> CONNECTED
        WPeerConnectionState.DISCONNECTED -> DISCONNECTED
        WPeerConnectionState.FAILED -> FAILED
        WPeerConnectionState.CLOSED -> CLOSED
      }
    }
  }
}

File: ios/Classes/model/PeerConnectionState.swift
import WebRTC

/// Representation of an `RTCPeerConnectionState`.
enum PeerConnectionState: Int {
  /// Any of the ICE transports or DTLS transports are in the `new` state and
  /// none of the transports are in the `connecting`, `checking`, `failed` or
  /// `disconnected` state, or all transports are in the `closed` state, or
  /// there are no transports.
  case new

  /// Any of the ICE transports or DTLS transports are in the `connecting` or
  /// `checking` state and none of them is in the `failed` state.
  case connecting

  /// All the ICE transports and DTLS transports are in the `connected`,
  /// `completed` or `closed` state, and at least one of them is in the
  /// `connected` or `completed` state.
  case connected

  /// Any of the ICE transports or DTLS transports are in the `disconnected`
  /// state, and none of them are in the `failed`, `connecting` or `checking`
  /// state.
  case disconnected

  /// Any of the ICE transports or DTLS transports are in the `failed` state.
  case failed

  /// Peer connection is closed.
  case closed

  /// Converts the provided `RTCPeerConnectionState` into a
  /// `PeerConnectionState`.
  static func fromWebRtc(state: RTCPeerConnectionState) -> PeerConnectionState {
    switch state {
    case .new:
      return PeerConnectionState.new
    case .connecting:
      return PeerConnectionState.connecting
    case .connected:
      return PeerConnectionState.connected
    case .disconnected:
      return PeerConnectionState.disconnected
    case .failed:
      return PeerConnectionState.failed
    case .closed:
      return PeerConnectionState.closed
    }
  }
}
